lib/common/tests/rules/pcmk__evaluate_date_expression_test.c

```diff
@@ -0,0 +1,1061 @@
+    crm_time_free(t);
+
+    // t is before start, next change is between start and end
+    t = crm_time_new("2024-02-01 11:59:59");
+    next_change = crm_time_new("2024-02-01 14:00:00");
+    reference = crm_time_new("2024-02-01 12:00:00");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_before_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is equal to start, next change is between start and end
+    t = crm_time_new("2024-02-01 12:00:00");
+    next_change = crm_time_new("2024-02-01 14:30:00");
+    reference = crm_time_new("2024-02-01 14:30:00");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_within_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is between start and end, next change is after end
+    t = crm_time_new("2024-02-01 14:03:11");
+    next_change = crm_time_new("2024-04-04 04:04:04");
+    reference = crm_time_new("2024-02-01 15:00:01");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_within_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is equal to end, next change is after end
+    t = crm_time_new("2024-02-01 15:00:00");
+    next_change = crm_time_new("2028-04-04 04:04:04");
+    reference = crm_time_new("2024-02-01 15:00:01");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_within_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is after end, next change is after end
+    t = crm_time_new("2024-02-01 15:00:01");
+    next_change = crm_time_new("2028-04-04 04:04:04");
+    reference = crm_time_new("2028-04-04 04:04:04");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_after_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    free_xml(xml);
+}
+
+#define EXPR_RANGE_VALID_START_DURATION_MISSING_ID      \
+    "<" PCMK_XE_DATE_EXPRESSION " "                     \
+    PCMK_XA_OPERATION "='" PCMK_VALUE_IN_RANGE "' "     \
+    PCMK_XA_START "='2024-02-01 12:00:00'>"             \
+    "<" PCMK_XE_DURATION " " PCMK_XA_ID "='d' "         \
+    PCMK_XA_HOURS "='3' />"                             \
+    "</" PCMK_XE_DATE_EXPRESSION ">"
+
+static void
+range_valid_start_duration_missing_id(void **state)
+{
+    // Currently acceptable
```

Comment:
Could note explicitly that it should match the `range_valid_start_valid_duration` test block currently

Reply:
Meh, it's going to be a syntax error soon